### PR TITLE
upgrade syntax and configs to  allow terraform 0.12 usage

### PR DESCRIPTION
**Pull request by ezalejski**

### CODE REVIEW

File: modules/subscription/providers.tf [hunk 1]
```diff
@@ -0,0 +1,4 @@
+provider "google" {
```
Comment (rokerkony) [inline modules/subscription/providers.tf:1]: We didn't want to have a provider hardcoded here because you could use old version with old provider with a subset of fields... you can see that in the readme.

I would be ok of dropping support for this with a switch to tf12 but we need to change also the readme. I would remove a note which is there and write the min. version of google provider in compatibility section

File: variables.tf [hunk 2]
```diff
@@ -40,11 +50,14 @@ variable "definition" {
           - retentionDuration (optional) (default: 604800)
           - retainAcked (optional) (default: false)
           - expirationTtl (optional) (default: 2678400)
-  EOF
+  
+EOF
+
```
Comment (rokerkony) [inline variables.tf:55]: why this change? Indent seems to be weird and empty lines unnecessary, same few lines above, of course, if this is caused by tf fmt it is ok, but it doesn't look like...

File: versions.tf [hunk 3]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (rokerkony) [inline versions.tf:1]: empty line not needed, same in  subscription's versions.tf

**Review by rokerkony — CHANGES_REQUESTED** [review]

File: variables.tf (hunk 2 above)
Comment (ezalejski) [reply variables.tf:55]: I think vscode added this, I will see if I can fix it.

File: modules/subscription/providers.tf (hunk 1 above)
Comment (ezalejski) [reply modules/subscription/providers.tf]: 👍

File: modules/subscription/main.tf [hunk 4]
```diff
@@ -1,36 +1,53 @@
...
-  message_retention_duration = "${lookup(var.pullSubscriptions[count.index], "retentionDuration", "604800")}s"
-  retain_acked_messages = "${lookup(var.pullSubscriptions[count.index], "retainAcked", false)}"
-  expiration_policy = {
-    ttl = "${lookup(var.pullSubscriptions[count.index], "expirationPolicy", "2678400")}s"
+  name                 = var.pullSubscriptions[count.index]["name"]
+  ack_deadline_seconds = lookup(var.pullSubscriptions[count.index], "deadline", 10)
+  message_retention_duration = "${lookup(
+    var.pullSubscriptions[count.index],
+    "retentionDuration",
+    "604800",
```
Comment (filiphaftek) [inline modules/subscription/main.tf:16]: I would make this not hardocded if possible?

File: modules/subscription/main.tf [hunk 5]
```diff
@@ -1,36 +1,53 @@
...
+  )}s"
+  retain_acked_messages = lookup(var.pullSubscriptions[count.index], "retainAcked", false)
+  expiration_policy {
+    ttl = "${lookup(
+      var.pullSubscriptions[count.index],
+      "expirationPolicy",
+      "2678400",
```
Comment (filiphaftek) [inline modules/subscription/main.tf:23]: Same here.

File: modules/subscription/main.tf [hunk 6]
```diff
@@ -1,36 +1,53 @@
...
-  message_retention_duration = "${lookup(var.pushSubscriptions[count.index], "retentionDuration", "604800")}s"
-  retain_acked_messages = "${lookup(var.pushSubscriptions[count.index], "retainAcked", false)}"
-  expiration_policy = {
-    ttl = "${lookup(var.pushSubscriptions[count.index], "expirationPolicy", "2678400")}s"
+  name                 = var.pushSubscriptions[count.index]["name"]
+  ack_deadline_seconds = lookup(var.pushSubscriptions[count.index], "deadline", 10)
+  message_retention_duration = "${lookup(
+    var.pushSubscriptions[count.index],
+    "retentionDuration",
+    "604800",
```
Comment (filiphaftek) [inline modules/subscription/main.tf:38]: Same here.

File: modules/subscription/main.tf [hunk 7]
```diff
@@ -1,36 +1,53 @@
...
+  )}s"
+  retain_acked_messages = lookup(var.pushSubscriptions[count.index], "retainAcked", false)
+  expiration_policy {
+    ttl = "${lookup(
+      var.pushSubscriptions[count.index],
+      "expirationPolicy",
+      "2678400",
```
Comment (filiphaftek) [inline modules/subscription/main.tf:45]: Same here.

File: README.md [hunk 8]
```diff
@@ -8,6 +8,9 @@ These types of resources are supported:
 * [Subscription](https://www.terraform.io/docs/providers/google/r/pubsub_subscription.html)
 * [Topic IAM member](https://www.terraform.io/docs/providers/google/r/pubsub_topic_iam.html)
 
+# Compatibility
+This module is meant for use with Terraform 0.12. If you haven't upgraded and need a Terraform 0.11.x-compatible version of this module, the last released version intended for Terraform 0.11.x is 2.0.0.
+
 ## Note
 
 If you wish to use `retentionDuration`, `retainAcked` or `expirationTtl` you need the `google` provider `~> 2.20`
```
Comment (filiphaftek) [inline README.md:16]: I would add version in the example in the README, I guess it's v 3.0.0 as this is breaking change b/c requires new tf?

**Review by filiphaftek — CHANGES_REQUESTED** [review]

File: README.md (hunk 8 above)
Comment (ezalejski) [reply README.md]: I agree 👍

File: modules/subscription/main.tf (hunk 4 above)
Comment (ezalejski) [reply modules/subscription/main.tf:16]: Do you mean by dynamic block?  We are hardcoding only a default value. Am I correct @rokerkony ?

File: modules/subscription/main.tf (hunk 5 above)
Comment (ezalejski) [reply modules/subscription/main.tf:23]: As above :)

File: modules/subscription/main.tf (hunk 4 above)
Comment (rokerkony) [reply modules/subscription/main.tf:16]: yes, those are only default values if none is provided

**Comment by rokerkony** [comment]
you removed `outputs.tf` and `modules/subscription/outputs.tf` but I think it should be available to have it correct based on the guidelines:
> `main.tf`, `variables.tf`, `outputs.tf`. These are the recommended filenames for a minimal module, even if they’re empty.

see more: https://www.terraform.io/docs/modules/index.html

**Review by rokerkony — APPROVED** [review]

File: modules/subscription/main.tf (hunk 4 above)
Comment (ezalejski) [reply modules/subscription/main.tf:16]: 👍

**Review by filiphaftek — APPROVED** [review]
👍
